src/features/session_trace/aggregate/trace/storage.js

```diff
@@ -0,0 +1,283 @@
+import { globalScope } from '../../../../common/constants/runtime'
+import { MODE } from '../../../../common/session/constants'
+import { now } from '../../../../common/timing/now'
+import { parseUrl } from '../../../../common/url/parse-url'
+import { TraceNode } from './node'
+
+const ERROR_MODE_SECONDS_WINDOW = 30 * 1000 // sliding window of nodes to track when simply monitoring (but not harvesting) in error mode
+const SUPPORTS_PERFORMANCE_OBSERVER = 'PerformanceObserver' in globalScope && typeof globalScope.PerformanceObserver === 'function'
+
+const ignoredEvents = {
+  // we find that certain events make the data too noisy to be useful
+  global: { mouseup: true, mousedown: true },
+  // certain events are present both in the window and in PVT metrics.  PVT metrics are prefered so the window events should be ignored
+  window: { load: true, pagehide: true },
+  // when ajax instrumentation is disabled, all XMLHttpRequest events will return with origin = xhrOriginMissing and should be ignored
+  xhrOriginMissing: { ignoreAll: true }
+}
+const toAggregate = {
+  typing: [1000, 2000],
+  scrolling: [100, 1000],
+  mousing: [1000, 2000],
+  touching: [1000, 2000]
+}
+
+/** The purpose of this class is to manage, normalize, and retrieve ST nodes as needed without polluting the main ST modules */
+export class TraceStorage {
+  nodeCount = 0
+  trace = {}
+  earliestTimeStamp = Infinity
+  latestTimeStamp = 0
+
+  constructor (parent) {
+    this.parent = parent
+  }
+
+  /** Central function called by all the other store__ & addToTrace API to append a trace node. */
+  storeSTN (stn) {
+    if (this.parent.blocked) return
+    if (this.nodeCount >= this.parent.maxNodesPerHarvest) { // limit the amount of pending data awaiting next harvest
+      if (this.parent.agentRuntime.session.state.sessionTraceMode !== MODE.ERROR) {
+        // trigger a harvest early
+        this.parent.scheduler.runHarvest({ needResponse: true })
```

Comment:
in the run-through I did mention the harvest early change instead of dropping data. Dropping data is not something we should do (period) if we are now monetizing ST, there's literally no good reason to do so.

Reply:
`I don't like this as we have no idea what the status of the scheduler or trace aggregate is to be executing an initial harvest--or any harvest`

I'm not following the logic here.  It checks if it's in error mode in this block.  If the feature starts at any time in full mode, it immediately harvests... meaning error mode doesnt force a harvest, and full mode wouldve already harvested the "initial" harvest (which has no real meaning at the moment by the way)... This also has precedent as the SR feature does the same thing.

If it makes you feel better we can add a check for the harvester's status, or backlog the data in some way, but i think the risk here is pretty low considering the alternative you are proposing is to drop the data altogether

Reply:
I added a check for harvesting as well as a temp bucket to hold any 'blocked' data while harvesting

Reply:
`agentRuntime.session.state.sessionTraceMode` is something that can exist as long as session entity started up.
It doesn't give a clear indicator that the `scheduler` is ready, has been ready, or has gone through ready-shutdown process, etc.
In addition, again this is conflating what this class & function is meant to do. So I still think if nodes are supposed to be sent when buffered is full and trace is in FULL mode, that should be done in the aggregate/index.js, who also has the full visibility scope of scheduler.